clib/helpstr.c

```diff
@@ -43,115 +125,107 @@ int *ierr;
-    } else if(decloc != NULL) {
-      ch1=*(decloc+1);
-      ch2=*(decloc+2);
-      ch3=*(decloc+3);
-      if((ch1== '_' ||
-	  (ch1 == '3' &&  K4K3  == *rack) ||
-	  (ch1 == 'm' &&  MK3   == *rack) ||
-	  (ch1 == 'n' && (MK3   == *rack || MK4   == *rack ||
-	                  LBA4  == *rack)) ||
-	  (ch1 == 'e' && (MK3   == *rack || MK4   == *rack ||
-			  VLBA  == *rack || VLBA4 == *rack ||
-	                  LBA4  == *rack || DBBC  == *rack)) ||
-	  (ch1 == 'f' && (MK3   == *rack || MK4   == *rack ||
-			  K4K3  == *rack || K4MK4 == *rack ||
-			  K4    == *rack || LBA4  == *rack)) ||
-	  (ch1 == '4' &&  MK4   == *rack) ||
-	  (ch1 == 's' &&  S2    == *rack) ||
-	  (ch1 == 'g' && (MK4   == *rack || VLBA  == *rack ||
-			  VLBA4 == *rack || K4MK4 == *rack ||
-			  LBA4  == *rack)) ||
-	  (ch1 == 'h' && (MK4   == *rack || VLBA4 == *rack ||
-			  K4MK4 == *rack || LBA4  == *rack)) ||
-	  (ch1 == 'i' && (MK4   == *rack || VLBA  == *rack ||
-			  VLBA4 == *rack || K4MK4 == *rack)) ||
-	  (ch1 == 'v' && (VLBA  == *rack)) ||
-	  (ch1 == 'w' && (VLBA  == *rack || VLBA4 == *rack)) ||
-	  (ch1 == 'k' && (K4K3  == *rack || K4MK4 == *rack ||
-			  K4    == *rack)) ||
-	  (ch1 == 'l' && (LBA   == *rack || LBA4 == *rack)) ||
-	  (ch1 == 'd' &&  DBBC  == *rack) ||
-	  (ch1 == 'a' &&  0     != *rack)) &&
-	 (ch2 == '_' || ch2 == '+' ||
-	  (ch2 == 'k' &&  K4    == *drive1) ||
-	  (ch2 == 'm' &&  MK3   == *drive1) ||
-	  (ch2 == 'n' && (MK3   == *drive1 || MK4 == *drive1)) ||
-	  (ch2 == '4' &&  MK4   == *drive1) ||
-	  (ch2 == 's' &&  S2    == *drive1) ||
-	  (ch2 == 'w' && (VLBA  == *drive1 || VLBA4 == *drive1)) ||
-	  (ch2 == 'a' &&  0     != *drive1) ||
-	  (ch2 == 'l' && (MK3   == *drive1 || MK4   == *drive1 ||
-			  VLBA  == *drive1 || VLBA4 == *drive1 ))) &&
-	 (ch3 == '_' || ch3 == '+' ||
-	  (ch3 == 'k' &&  K4    == *drive2) ||
-	  (ch3 == 'm' &&  MK3   == *drive2) ||
-	  (ch3 == 'n' && (MK3   == *drive2 || MK4 == *drive2)) ||
-	  (ch3 == '4' &&  MK4   == *drive2) ||
-	  (ch3 == 's' &&  S2    == *drive2) ||
-	  (ch3 == 'w' && (VLBA  == *drive2 || VLBA4 == *drive2)) ||
-	  (ch3 == 'a' &&  0     != *drive2) ||
-	  (ch3 == 'l' && (MK3   == *drive2 || MK4   == *drive2 ||
-			  VLBA  == *drive2 || VLBA4 == *drive2 )))
-	 &&
-	 ((*drive1 !=0 && *drive2 !=0 && ((ch2 == '+' || ch3 == '+') ||
-					  (ch2 =='_' && ch3 == '_'))) ||
-	  ((*drive1 ==0 || *drive2 == 0) && (ch2 != '+' && ch3 != '+')))
-	 ) {
-        strcpy(runstr,outbuf);
-        *ierr = 0;
-        break;
-      }
-    }
-  }
-  fclose(idum);
+	char name[MAX_STRING] = {};
+	char *decloc;
+	char ch1, ch2, ch3;
+
+	*ierr = -3;
+
+	if (*clength > MAX_STRING) {
+		*ierr = -2;
+		return;
+	}
+
+	strncpy(name, cnam, *clength);
+	name[*clength] = '\0';
+
+	decloc = strchr(name, '.');
+	if (decloc != NULL) {
+		char *p = check_help_dirs(name);
+		if (p == NULL) {
+			return;
+		}
+		*ierr = 0;
+		strcpy(runstr, p);
+		free(p);
+		return;
+	}
+
+	char **paths = ls_prefix(help_dirs, name);
```

Comment:
I noticed that **help=vc**, apparently picks whatever **vcxx** it happens to hit first in the directory. There can also be less related prefix matches. Unless you intend this behavior I think it can be prevented adding

strcat(name,".",1); 

just before this line. Or to be "safe" maybe something ugly like:

strncat(name,".",sizeof(name)-strlen(name) < 2 ? 0 : 1);

Reply:
Good point

Reply:
OK, I just made sure the stack buffer was big enough to fit the null byte and the potential '.'

I'm always wary of the `strn*` functions; they never seem to do quite what I expect them to do and I have to think to hard when I re-read the code.